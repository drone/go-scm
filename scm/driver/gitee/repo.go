// Copyright 2017 Drone.IO Inc. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package gitee

import (
	"context"
	"fmt"
	"strconv"
	"time"

	"github.com/drone/go-scm/scm"
)

type RepositoryService struct {
	client *wrapper
}

func (s *RepositoryService) Find(ctx context.Context, repo string) (*scm.Repository, *scm.Response, error) {
	path := fmt.Sprintf("repos/%s", repo)
	out := new(repository)
	res, err := s.client.do(ctx, "GET", path, nil, out)
	return convertRepository(out), res, err
}

func (s *RepositoryService) FindHook(ctx context.Context, repo string, id string) (*scm.Hook, *scm.Response, error) {
	path := fmt.Sprintf("repos/%s/hooks/%s", repo, id)
	out := new(hook)
	res, err := s.client.do(ctx, "GET", path, nil, out)
	return convertHook(out), res, err
}

func (s *RepositoryService) FindPerms(ctx context.Context, repo string) (*scm.Perm, *scm.Response, error) {
	repos, res, err := s.Find(ctx, repo)
	if err == nil {
		return repos.Perm, res, err
	}
	return nil, res, err
}

func (s *RepositoryService) List(ctx context.Context, opts scm.ListOptions) ([]*scm.Repository, *scm.Response, error) {
	path := fmt.Sprintf("user/repos?%s", encodeListOptions(opts))
	out := []*repository{}
	res, err := s.client.do(ctx, "GET", path, nil, &out)
	return convertRepositoryList(out), res, err
}

func (s *RepositoryService) ListHooks(ctx context.Context, repo string, opts scm.ListOptions) ([]*scm.Hook, *scm.Response, error) {
	path := fmt.Sprintf("repos/%s/hooks?%s", repo, encodeListOptions(opts))
	out := []*hook{}
	res, err := s.client.do(ctx, "GET", path, nil, &out)
	return convertHookList(out), res, err
}

func (s *RepositoryService) ListStatus(context.Context, string, string, scm.ListOptions) ([]*scm.Status, *scm.Response, error) {
	return nil, nil, scm.ErrNotSupported
}

func (s *RepositoryService) CreateHook(ctx context.Context, repo string, input *scm.HookInput) (*scm.Hook, *scm.Response, error) {
	path := fmt.Sprintf("repos/%s/hooks", repo)
	in := new(hook)
	// 1: signature
	in.EncryptionType = 1
	in.Password = input.Secret
	in.URL = input.Target
	convertFromHookEvents(input.Events, in)

	out := new(hook)
	res, err := s.client.do(ctx, "POST", path, in, out)
	return convertHook(out), res, err
}

func (s *RepositoryService) CreateStatus(context.Context, string, string, *scm.StatusInput) (*scm.Status, *scm.Response, error) {
	return nil, nil, scm.ErrNotSupported
}

func (s *RepositoryService) UpdateHook(ctx context.Context, repo, id string, input *scm.HookInput) (*scm.Hook, *scm.Response, error) {
	path := fmt.Sprintf("repos/%s/hooks/%s", repo, id)
	in := new(hook)
	// 1: signature
	in.EncryptionType = 1
	in.Password = input.Secret
	in.URL = input.Target
	convertFromHookEvents(input.Events, in)

	out := new(hook)
	res, err := s.client.do(ctx, "PATCH", path, in, out)
	return convertHook(out), res, err
}

func (s *RepositoryService) DeleteHook(ctx context.Context, repo, id string) (*scm.Response, error) {
	path := fmt.Sprintf("repos/%s/hooks/%s", repo, id)
	return s.client.do(ctx, "DELETE", path, nil, nil)
}

type repository struct {
	ID    int `json:"id"`
	Owner struct {
		ID        int    `json:"id"`
		Login     string `json:"login"`
		Name      string `json:"name"`
		AvatarURL string `json:"avatar_url"`
	} `json:"owner"`
	Namespace     namespace `json:"namespace"`
	Name          string    `json:"name"`
	FullName      string    `json:"full_name"`
	HumanName     string    `json:"human_name"`
	Path          string    `json:"path"`
	Public        bool      `json:"public"`
	Private       bool      `json:"private"`
	Internal      bool      `json:"internal"`
	Fork          bool      `json:"fork"`
	URL           string    `json:"url"`
	HtmlURL       string    `json:"html_url"`
	SshURL        string    `json:"ssh_url"`
	DefaultBranch string    `json:"default_branch"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`
	Permission    struct {
		Admin bool `json:"admin"`
		Push  bool `json:"push"`
		Pull  bool `json:"pull"`
	} `json:"permission"`
}

type hook struct {
	ID        int       `json:"id"`
	URL       string    `json:"url"`
	CreatedAt time.Time `json:"created_at"`
	// password encryption type, 0: password, 1: signature
	EncryptionType      int    `json:"encryption_type"`
	Password            string `json:"password"`
	ProjectID           int    `json:"project_id"`
	Result              string `json:"result"`
	ResultCode          int    `json:"result_code"`
	PushEvents          bool   `json:"push_events"`
	TagPushEvents       bool   `json:"tag_push_events"`
	IssuesEvents        bool   `json:"issues_events"`
	NoteEvents          bool   `json:"note_events"`
	MergeRequestsEvents bool   `json:"merge_requests_events"`
}

type namespace struct {
	ID      int    `json:"id"`
	Type    string `json:"type"`
	Name    string `json:"name"`
	Path    string `json:"path"`
	HtmlURL string `json:"html_url"`
}

func convertRepositoryList(from []*repository) []*scm.Repository {
	to := []*scm.Repository{}
	for _, v := range from {
		to = append(to, convertRepository(v))
	}
	return to
}
func convertRepository(from *repository) *scm.Repository {
	return &scm.Repository{
		ID:        strconv.Itoa(from.ID),
		Name:      from.Name,
		Namespace: from.Namespace.Path,
		Perm: &scm.Perm{
			Push:  from.Permission.Push,
			Pull:  from.Permission.Pull,
			Admin: from.Permission.Admin,
		},
		Link:     from.HtmlURL,
		Branch:   from.DefaultBranch,
		Private:  from.Private,
		Clone:    from.HtmlURL,
		CloneSSH: from.SshURL,
		Created:  from.CreatedAt,
		Updated:  from.UpdatedAt,
	}
}

func convertHookList(from []*hook) []*scm.Hook {
	to := []*scm.Hook{}
	for _, v := range from {
		to = append(to, convertHook(v))
	}
	return to
}
func convertHook(from *hook) *scm.Hook {
	return &scm.Hook{
		ID:         strconv.Itoa(from.ID),
		Active:     true,
		Target:     from.URL,
		Events:     convertHookEvent(from),
		SkipVerify: true,
	}
}

func convertHookEvent(from *hook) []string {
	var events []string
	if from.PushEvents {
		events = append(events, "push")
	}
	if from.TagPushEvents {
		events = append(events, "tag_push")
	}
	if from.IssuesEvents {
		events = append(events, "issues")
	}
	if from.NoteEvents {
		events = append(events, "note")
	}
	if from.MergeRequestsEvents {
		events = append(events, "merge_requests")
	}
	return events
}

// convertFromHookEvents not support: Branch, Deployment
func convertFromHookEvents(from scm.HookEvents, to *hook) {
	if from.Push {
		to.PushEvents = true
	}
	if from.PullRequest {
		to.MergeRequestsEvents = true
	}
	if from.Issue {
		to.IssuesEvents = true
	}
	if from.IssueComment || from.PullRequestComment || from.ReviewComment {
		to.NoteEvents = true
	}
	if from.Tag {
		to.TagPushEvents = true
	}
}
