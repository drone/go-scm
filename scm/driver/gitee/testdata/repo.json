{
  "id": 14836026,
  "full_name": "kit101/drone-yml-test",
  "human_name": "kit101/drone-yml-test",
  "url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test",
  "namespace": {
    "id": 1576684,
    "type": "personal",
    "name": "kit101",
    "path": "kit101",
    "html_url": "https://gitee.com/kit101"
  },
  "path": "drone-yml-test",
  "name": "drone-yml-test",
  "owner": {
    "id": 1535738,
    "login": "kit101",
    "name": "kit101",
    "avatar_url": "https://portrait.gitee.com/uploads/avatars/user/511/1535738_qkssk1711_1578953939.png",
    "url": "https://gitee.com/api/v5/users/kit101",
    "html_url": "https://gitee.com/kit101",
    "remark": "",
    "followers_url": "https://gitee.com/api/v5/users/kit101/followers",
    "following_url": "https://gitee.com/api/v5/users/kit101/following_url{/other_user}",
    "gists_url": "https://gitee.com/api/v5/users/kit101/gists{/gist_id}",
    "starred_url": "https://gitee.com/api/v5/users/kit101/starred{/owner}{/repo}",
    "subscriptions_url": "https://gitee.com/api/v5/users/kit101/subscriptions",
    "organizations_url": "https://gitee.com/api/v5/users/kit101/orgs",
    "repos_url": "https://gitee.com/api/v5/users/kit101/repos",
    "events_url": "https://gitee.com/api/v5/users/kit101/events{/privacy}",
    "received_events_url": "https://gitee.com/api/v5/users/kit101/received_events",
    "type": "User"
  },
  "assigner": {
    "id": 1535738,
    "login": "kit101",
    "name": "kit101",
    "avatar_url": "https://portrait.gitee.com/uploads/avatars/user/511/1535738_qkssk1711_1578953939.png",
    "url": "https://gitee.com/api/v5/users/kit101",
    "html_url": "https://gitee.com/kit101",
    "remark": "",
    "followers_url": "https://gitee.com/api/v5/users/kit101/followers",
    "following_url": "https://gitee.com/api/v5/users/kit101/following_url{/other_user}",
    "gists_url": "https://gitee.com/api/v5/users/kit101/gists{/gist_id}",
    "starred_url": "https://gitee.com/api/v5/users/kit101/starred{/owner}{/repo}",
    "subscriptions_url": "https://gitee.com/api/v5/users/kit101/subscriptions",
    "organizations_url": "https://gitee.com/api/v5/users/kit101/orgs",
    "repos_url": "https://gitee.com/api/v5/users/kit101/repos",
    "events_url": "https://gitee.com/api/v5/users/kit101/events{/privacy}",
    "received_events_url": "https://gitee.com/api/v5/users/kit101/received_events",
    "type": "User"
  },
  "description": "",
  "private": false,
  "public": true,
  "internal": false,
  "fork": false,
  "html_url": "https://gitee.com/kit101/drone-yml-test.git",
  "ssh_url": "git@gitee.com:kit101/drone-yml-test.git",
  "forks_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/forks",
  "keys_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/keys{/key_id}",
  "collaborators_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/collaborators{/collaborator}",
  "hooks_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/hooks",
  "branches_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/branches{/branch}",
  "tags_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/tags",
  "blobs_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/blobs{/sha}",
  "stargazers_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/stargazers",
  "contributors_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/contributors",
  "commits_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/commits{/sha}",
  "comments_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/comments{/number}",
  "issue_comment_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/issues/comments{/number}",
  "issues_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/issues{/number}",
  "pulls_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/pulls{/number}",
  "milestones_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/milestones{/number}",
  "notifications_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/notifications{?since,all,participating}",
  "labels_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/labels{/name}",
  "releases_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/releases{/id}",
  "recommend": false,
  "gvp": false,
  "homepage": "https://gitee.com/kit101/drone-yml-test",
  "language": null,
  "forks_count": 0,
  "stargazers_count": 0,
  "watchers_count": 1,
  "default_branch": "master",
  "open_issues_count": 2,
  "has_issues": true,
  "has_wiki": true,
  "issue_comment": false,
  "can_comment": true,
  "pull_requests_enabled": true,
  "has_page": false,
  "license": null,
  "outsourced": false,
  "project_creator": "kit101",
  "members": [
    "kit101"
  ],
  "pushed_at": "2021-09-30T18:10:23+08:00",
  "created_at": "2021-03-24T11:24:34+08:00",
  "updated_at": "2021-09-30T18:10:27+08:00",
  "parent": null,
  "paas": null,
  "stared": false,
  "watched": true,
  "permission": {
    "pull": true,
    "push": true,
    "admin": true
  },
  "relation": "master",
  "assignees_number": 1,
  "testers_number": 1,
  "assignee": [
    {
      "id": 1535738,
      "login": "kit101",
      "name": "kit101",
      "avatar_url": "https://portrait.gitee.com/uploads/avatars/user/511/1535738_qkssk1711_1578953939.png",
      "url": "https://gitee.com/api/v5/users/kit101",
      "html_url": "https://gitee.com/kit101",
      "remark": "",
      "followers_url": "https://gitee.com/api/v5/users/kit101/followers",
      "following_url": "https://gitee.com/api/v5/users/kit101/following_url{/other_user}",
      "gists_url": "https://gitee.com/api/v5/users/kit101/gists{/gist_id}",
      "starred_url": "https://gitee.com/api/v5/users/kit101/starred{/owner}{/repo}",
      "subscriptions_url": "https://gitee.com/api/v5/users/kit101/subscriptions",
      "organizations_url": "https://gitee.com/api/v5/users/kit101/orgs",
      "repos_url": "https://gitee.com/api/v5/users/kit101/repos",
      "events_url": "https://gitee.com/api/v5/users/kit101/events{/privacy}",
      "received_events_url": "https://gitee.com/api/v5/users/kit101/received_events",
      "type": "User"
    }
  ],
  "testers": [
    {
      "id": 1535738,
      "login": "kit101",
      "name": "kit101",
      "avatar_url": "https://portrait.gitee.com/uploads/avatars/user/511/1535738_qkssk1711_1578953939.png",
      "url": "https://gitee.com/api/v5/users/kit101",
      "html_url": "https://gitee.com/kit101",
      "remark": "",
      "followers_url": "https://gitee.com/api/v5/users/kit101/followers",
      "following_url": "https://gitee.com/api/v5/users/kit101/following_url{/other_user}",
      "gists_url": "https://gitee.com/api/v5/users/kit101/gists{/gist_id}",
      "starred_url": "https://gitee.com/api/v5/users/kit101/starred{/owner}{/repo}",
      "subscriptions_url": "https://gitee.com/api/v5/users/kit101/subscriptions",
      "organizations_url": "https://gitee.com/api/v5/users/kit101/orgs",
      "repos_url": "https://gitee.com/api/v5/users/kit101/repos",
      "events_url": "https://gitee.com/api/v5/users/kit101/events{/privacy}",
      "received_events_url": "https://gitee.com/api/v5/users/kit101/received_events",
      "type": "User"
    }
  ],
  "status": "开始",
  "empty_repo": false,
  "programs": [],
  "enterprise": null
}