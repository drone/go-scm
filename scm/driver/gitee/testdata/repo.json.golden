{
  "ID": "14836026",
  "Namespace": "kit101",
  "Name": "drone-yml-test",
  "Perm": {
    "Pull": true,
    "Push": true,
    "Admin": true
  },
  "Branch": "master",
  "Private": false,
  "Clone": "https://gitee.com/kit101/drone-yml-test.git",
  "CloneSSH": "git@gitee.com:kit101/drone-yml-test.git",
  "Link": "https://gitee.com/kit101/drone-yml-test.git",
  "Created": "2021-03-24T11:24:34+08:00",
  "Updated": "2021-09-30T18:10:27+08:00"
}